generic/shared/src/main/scala/monocle/generic/internal/CoproductToDisjunction.scala

```diff
@@ -0,0 +1,35 @@
+package monocle.generic.internal
+
+import scalaz.{\/, -\/, \/-}
+import shapeless._
+
+/**
+  * Typeclass converting a [[Coproduct]] to an [[\/]].
+  *
+  * Based on [[shapeless.ops.coproduct.CoproductToEither]]:
+  * https://github.com/milessabin/shapeless/blob/shapeless-2.3.3-scala-2.13.0-M4/core/src/main/scala/shapeless/ops/coproduct.scala#L1275-L1303
+  */
+sealed trait CoproductToDisjunction[C <: Coproduct] extends DepFn1[C] with Serializable
+
+object CoproductToDisjunction {
+  type Aux[In <: Coproduct, Out0] = CoproductToDisjunction[In] { type Out = Out0 }
+
+  implicit def baseToEither[L, R]: CoproductToDisjunction.Aux[L :+: R :+: CNil, L \/ R] = new CoproductToDisjunction[L :+: R :+: CNil] {
+    type Out = L \/ R
+    def apply(t: L :+: R :+: CNil): L \/ R = t match {
+      case Inl(l)      => -\/(l)
+      case Inr(Inl(r)) => \/-(r)
+      case _           => ???
```

Comment:
what does it mean?

Reply:
This code was taken [straight from shapeless][0], but the missing pattern is `Inr(Inr(c))`, where `c` is of type `CNil`.

`CNil` is an unhinhabited type, it's only used to signal the end of a `:+:` list, so this pattern is guaranteed to never match.

However, I think replacing `case _ => ???` with `case Inr(Inr(cnil)) => cnil.impossible` would be more readable.

`CNil` has this `def impossible: Nothing` member that we're supposed to use it when pattern matching on it:

```scala
/** The CNil type is used to terminate a 'list' of :+: alternatives.
  *
  * Like the Nil constructor of List, it does not convey real information.
  * This is achieved by not having any value for CNil.
  *
  * This makes the type `Int :+: CNil` equivalent to `Int`, because the right (`Inr`) alternative
  * of `:+:` can not be constructed properly.
  */
sealed trait CNil extends Coproduct {
  /** Call this when you hit the CNil case in pattern matching to make the match exhaustive and safe. */
  def impossible: Nothing
}
```

I'll make this change.


 [0]: https://github.com/milessabin/shapeless/blob/shapeless-2.3.3-scala-2.13.0-M4/core/src/main/scala/shapeless/ops/coproduct.scala#L1275-L1303